### ISSUE TITLE
Locally Aware Feature Comparison

### ISSUE THREAD

**Issue by NilsB98** [NilsB98/Diffusion-Based-AD#16]
To improve the anomaly map use not only a pixel-wise difference between the reconstruction and the original image, but include a feature based comparison with features extract by a second network.
See [this](https://arxiv.org/pdf/2305.15956v1.pdf) paper.

**Comment by Henry0528** [comment]
hello，are you still working on this project?

**Comment by NilsB98** [comment]
Hi Henry, absolutely! 
Just did the work locally recently.
If there is some specific functionalitiy that you'd wish for feel free to make proposals, I'm always glad to get some interaction and feedback :)

**Comment by Henry0528** [comment]
Thank you very much for your reply! I wonder whether you have some final results about the model performance of each class like image-level and pixel-level AUROC?

**Comment by NilsB98** [comment]
I'll have to train the model for the different objects, currently I got the recent metric only for the hazelnut data. 
There I'm able to achieve a pixel-level f1 of 0.711 and a img-level f1 of 0.942.

I'll try to run the experiments on the other MVTEC objects over the weekend as well and can give you the scores too, but also refactoring my local code s.t. it's useful and runnable by others with this repo.

**Comment by Henry0528** [comment]
Hello, I have been working on your code and trying to make some modifications, but I have met some problems. I changed the image input by the features extracted by a pre-trained EfficientNet-b4(272*32*32 in size) and used all normal images neglecting the classes to train a unified model that can detect all classes in the MVTec dataset. While I got extremely low-performance testing by the AUROC on both image-level and pixel-level. I have uploaded the code and I would be grateful if you could give any suggestions.

**Comment by NilsB98** [comment]
Hi, sure I can check it out, is there a specific file that you already suspect to not work properly? Feel free to link it.
Other than that I'm currently working on the feature extraction task as well and will push the code soon.

**Comment by Henry0528** [comment]
Here is the code [link](https://github.com/Henry0528/Diffusion-Based-AD), I have made some changes in main.py and used test.py to evaluate the model performance while getting a really bad result.  Here is the [paper ](https://arxiv.org/abs/2307.08059)I'm reproducing, I want to reproduce the baseline model performance in Tab.4(LafitE w/o F.E).

**Comment by NilsB98** [comment]
Okay, thanks, the approach mentioned in the paper does sound interesting indeed. I'll probably check on your code tomorrow evening and see whether I can find out where things might go wrong.

**Comment by Henry0528** [comment]
Hello, have you got any idea? I have tried many different settings but struggled to get a good baseline model.

**Comment by NilsB98** [comment]
Hi Henry, again sorry for the late response. 
Somehow github doesn't really show the differences between the files you uploaded and the original code quite well and just marks the whole files as different.
Did you already try to implement parts of the mentioned paper, or is it only the metrics that you added?

If you only changed the metrics, then I can provide you with some code which should work pretty much out of the box, and give good results.
Otherwise can you point me to the implementations that you made besides the metrics, if there are any?

**Comment by Henry0528** [comment]
I've made two modifications. 1. according to the mentioned paper, I used the feature extracted by an Efficientnet as input in main.py(line 156 and line 213-217)  2, in test.py, I added image and pixel AUCROC metrics in line 161. BTW could you add the  image and pixel AUCROC metrics to the original code?

**Comment by Henry0528** [comment]
In my opinion,  there are lots of methods to get the final 1-dimension anomaly map. The way you used is OK but I think the more common operation is to first calculate a diffmap between original images or features and the reconstructed ones with (H, W,C), then we get the final anomaly map by simply summing up or taking means of all channels like using torch.sum or torch.mean to get the (H,W,1) map. I think the different ways to calculate the anomaly map are not the key factors that affect the final results.

**Comment by NilsB98** [comment]
Yea I also think that this shouldn't be the decisive factor. I'll clone your repo and to some testing and debugging to see whether I might find out where things might go wrong or where we might have missed something else.
